makeflow/src/makeflow_module_archive.c

```diff
@@ -0,0 +1,472 @@
+ *            |                                                 |--> output_files --> file_name(symlink to actual file)
+ *            |--> files --> checksum_pre(2 digits) --> checksum (actual file)
+ */
+
+/* Write the task and run info to the task directory
+ *	These files are hardcoded to task_info and run_info */
+static int makeflow_archive_write_task_info(struct archive_instance *a, struct dag_node *n, struct batch_task *t, char *archive_path) {
+
+	struct batch_file *f;
+	char *task_info = string_format("%s/task_info", archive_path);
+
+	FILE *fp = fopen(task_info, "w");
+	if (fp == NULL) {
+		free(task_info);
+		debug(D_ERROR|D_MAKEFLOW_HOOK, "could not create task_info for node %d archive", n->nodeid);
+		return 0;
+	} else {
+/* task_info :
+ *	COMMAND: Tasks command that was run
+ *	SRC_COMMAND: Origin node's command for reference
+ *	SRC_LINE:  Line of origin node in SRC_MAKEFLOW
+ *	SRC_MAKEFLOW:  ID of file for the original Makeflow stored in archive
+ *	INPUT_FILES: Alphabetic list of input files checksum IDs
+ *	OUTPUT_FILES: Alphabetic list of output file outer_names
+ */
+		fprintf(fp, "COMMAND : %s\n", t->command);
+		fprintf(fp, "SRC_COMMAND : %s\n", n->command);
+		fprintf(fp, "SRC_LINE : %d\n", n->linenum);
+		fprintf(fp, "SRC_MAKEFLOW : %s\n", a->source_makeflow);
+		fprintf(fp, "INPUT_FILES : \n");
+		struct list_cursor *cur = list_cursor_create(t->input_files);
+		for(list_seek(cur, 0); list_get(cur, (void**)&f); list_next(cur)) {
+			char *id = batch_file_generate_id(f);
+			fprintf(fp, "\t%s : %s\n", f->outer_name, id);
+			free(id);
+		}
+		list_cursor_destroy(cur);
+
+		fprintf(fp, "OUTPUT_FILES : \n");
+		cur = list_cursor_create(t->output_files);
+		for(list_seek(cur, 0); list_get(cur, (void**)&f); list_next(cur)) {
+			char *id = batch_file_generate_id(f);
+			fprintf(fp, "\t%s : %s\n", f->outer_name, id);
+			free(id);
+		}
+		list_cursor_destroy(cur);
+	}
+	fclose(fp);
+	free(task_info);
+
+	task_info = string_format("%s/run_info", archive_path);
+
+	fp = fopen(task_info, "w");
+	if (fp == NULL) {
+		free(task_info);
+		debug(D_ERROR|D_MAKEFLOW_HOOK, "could not create run_info for node %d archive", n->nodeid);
+		return 0;
+	} else {
+/* run_info : 
+ *  SUBMITTED : Time task was submitted
+ *  STARTED : Time task was started
+ *  FINISHED : Time task was completed
+ *  EXIT_NORMALLY : 0 if abnormal exit, 1 is normal
+ *  EXIT_CODE : Task's exit code
+ *  EXIT_SIGNAL : Int value of signal if occurred
+ */
+		fprintf(fp, "SUBMITTED : %lu\n", t->info->submitted);
+		fprintf(fp, "STARTED : %lu\n", t->info->started);
+		fprintf(fp, "FINISHED : %lu\n", t->info->finished);
+		fprintf(fp, "EXIT_NORMAL : %d\n", t->info->exited_normally);
+		fprintf(fp, "EXIT_CODE : %d\n", t->info->exit_code);
+		fprintf(fp, "EXIT_SIGNAL : %d\n", t->info->exit_signal);
+	}
+	fclose(fp);
+	free(task_info);
+
+	return 1;
+}
+
+/* Archive the specified file.
```

Comment:
If there is an error or failure while archiving a file/task, will it corrupt the archive?

Reply:
Thats a good point. Currently it will be an incomplete/corrupt entry. I will think about removing partial entries on failure.